Be compatible with Java 16

I encountered several issues when trying to build ControlsFX using OpenJDK 16 and OpenJFX 16. Mainly because OpenJDK 16 has an issue with the gradle version we are using, and if I upgraded Gradle, I had issues with the gradle script itself.

I will let my investigation here in case anyone has the same issue, and that it might be useful in the future. I'm planning on migrating my enterprise application from Java 8 to Java 17 therefore using ControlsFX.

-  Building with OpenJDK 16 using gradle wrapper (6.0.1) throws this exception:
```
Could not open cp_init remapped class cache for 86qn5sz6lnn8dinx4xyta0vad (C:\Users\\.gradle\caches\6.0.1\scripts-remapped\nb_tooling_6ukhmbjlntsqsc0ggimovv2f\86qn5sz6lnn8dinx4xyta0vad\cp_init27b7120d8bed45e60dd0b95b8e5ecea0).
> Could not open cp_init generic class cache for initialization script 'C:\Program Files\NetBeans-12.3\netbeans\extide\modules\gradle\nb-tooling.gradle' (C:\Users\\.gradle\caches\6.0.1\scripts\86qn5sz6lnn8dinx4xyta0vad\cp_init\cp_init27b7120d8bed45e60dd0b95b8e5ecea0).
   > Could not initialize class org.codehaus.groovy.vmplugin.v7.Java7
```
It seems to be linked to the version of Gradle : https://stackoverflow.com/questions/61289461/java-lang-noclassdeffounderror-could-not-initialize-class-org-codehaus-groovy-v

- Upgrading the Gradle version to **6.8.2** for example, now triggers this exception:
```
Execution failed for task ':controlsfx:jar'.
> java.util.ConcurrentModificationException (no error message)
```
Somehow, the issue could be caused by the version of **biz.aQute.bnd.builder** (see https://github.com/gradle/gradle/issues/13628#issuecomment-672541910) so I've pushed this from **5.0.0** to **5.1.2**.

- But now I have 
```
Execution failed for task ':controlsfx:javadoc'.
> class java.lang.String cannot be cast to class java.util.List (java.lang.String and java.util.List are in module java.base of loader 'bootstrap')
```
I did not know what to do with that, so I simply remove the **javadoc** commands in the gradle scripts. Without that, it successfully builds.

- Pushing to Gradle **7.0.2** triggers these exception:
```
> Plugin with id 'maven' not found.
```
```
A problem occurred evaluating project ':controlsfx'.
> Could not find method testCompile() for arguments [junit:junit:[4,)] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.
```

Linked to #1287

Wow Gradle is a mess. I have wasted more than a few hours figuring things out and it seems that Gradle is performing some magic which is causing `controlsfx-samples` to fail as an application. I crossed check the `java` command and it seems fine.

With Gradle 7.x, JPMS support is auto enabled if a `module-info.java` is found in a project. This causing builds to fail.

I tried to switch it off using without any success:

```
java {
    modularity.inferModulePath = false
}
```

The other problem with our setup is that mainclass is present in `org.controlsfx.fxsampler` module and not in org.controlsfx.samples`. However, Gradle doesn't respect its own configuration. Using

```
application {
    mainModule.set('org.controlsfx.fxsampler')
    mainClass.set('fxsampler.FXSampler')
}
```

the Java command executed while running `gradlew run` is:

```
java ... --module org.controlsfx.samples/fxsampler.FXSampler
```

I tried to manually copy the java command and change the module, however, I kept getting:

```
> Task :controlsfx-samples:run FAILED
Error occurred during initialization of boot layer
java.lang.module.FindException: Error reading module: controlsfx/controlsfx-samples/build/classes/java/main
Caused by: java.lang.module.InvalidModuleDescriptorException: Package fxsampler not found in module
```

This issue doesn't just occur with Gradle 7. Try updating to Gradle 6.4 breaks the entire build. Gradle 6.4 doesn't enable JPMS automatically and is only supposed to have on-demand support for JPMS.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -11,11 +11,6 @@ jobs:
   verify:
     runs-on: ubuntu-latest
     steps:
-      - name: Install packages (Linux)
-        if: runner.os == 'Linux'
-        run: |
-          sudo apt-get install xvfb tigervnc-standalone-server tigervnc-common
-
       - name: Checkout
         uses: actions/checkout@v2
 
@@ -33,12 +28,8 @@ jobs:
         if: runner.os == 'Linux'
         run: |
           export DISPLAY=:90
-          mkdir /home/runner/.vnc
-          echo 123456 | vncpasswd -f > /home/runner/.vnc/passwd
-          chmod -v 600 /home/runner/.vnc/passwd
-          vncserver :90 -localhost -nolisten tcp
+          Xvfb -ac :90 -screen 0 1280x1024x24 > /dev/null 2>&1 &
           ./gradlew build test
-          vncserver -kill :90
 
       - name: Release Snapshots
         if: github.ref == 'refs/heads/master'

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -8,11 +8,6 @@ jobs:
   release:
     runs-on: ubuntu-latest
     steps:
-      - name: Install packages (Linux)
-        if: runner.os == 'Linux'
-        run: |
-          sudo apt-get install xvfb tigervnc-standalone-server tigervnc-common
-
       - name: Checkout
         uses: actions/checkout@v2
         with:
@@ -42,12 +37,8 @@ jobs:
         if: runner.os == 'Linux'
         run: |
           export DISPLAY=:90
-          mkdir /home/runner/.vnc
-          echo 123456 | vncpasswd -f > /home/runner/.vnc/passwd
-          chmod -v 600 /home/runner/.vnc/passwd
-          vncserver :90 -localhost -nolisten tcp
+          Xvfb -ac :90 -screen 0 1280x1024x24 > /dev/null 2>&1 &
           ./gradlew build test
-          vncserver -kill :90
 
       - name: Release to Maven Central
         id: deploy

--- a/build.gradle
+++ b/build.gradle
@@ -21,7 +21,7 @@ ext.java9RuntimeArgs = [
 subprojects {
 
     apply plugin: 'java'
-    apply plugin: 'maven'
+    apply plugin: 'maven-publish'
     apply plugin: 'org.openjfx.javafxplugin'
     apply from  : '../mavenPublish.gradle'
 

--- a/controlsfx-samples/build.gradle
+++ b/controlsfx-samples/build.gradle
@@ -2,8 +2,10 @@ ext.moduleName = 'org.controlsfx.samples'
 
 apply plugin: 'application'
 
-applicationName = 'FXSampler'
-mainClassName = 'org.controlsfx.fxsampler/fxsampler.FXSampler'
+application {
+    mainModule = 'org.controlsfx.samples'
+    mainClass = 'org.controlsfx.ControlsFXSampler'
+}
 
 version = controlsfx_version
 
@@ -58,7 +60,7 @@ jar {
         attributes  'Implementation-Title':     'ControlsFX-Samples',
                     'Implementation-Version':   project.version,
                     'Class-Path':               configurations.compileClasspath.collect { it.getName() }.join(' '),
-                    'Main-Class':               mainClassName
+                    'Main-Class':               application.mainClass
     }
 }
 

--- a/controlsfx-samples/src/main/java/module-info.java
+++ b/controlsfx-samples/src/main/java/module-info.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2018, 2020 ControlsFX
+ * Copyright (c) 2018, 2022, ControlsFX
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -52,5 +52,5 @@ module org.controlsfx.samples {
     opens org.controlsfx.samples.tableview to javafx.base;
     opens org.controlsfx.samples.spreadsheet to javafx.graphics;
     
-    provides fxsampler.FXSamplerProject with org.controlsfx.ControlsFXSampler;
+    provides fxsampler.FXSamplerProject with org.controlsfx.ControlsFXSamplerProject;
 }

--- a/controlsfx-samples/src/main/java/org/controlsfx/ControlsFXSampler.java
+++ b/controlsfx-samples/src/main/java/org/controlsfx/ControlsFXSampler.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2013, 2020, ControlsFX
+ * Copyright (c) 2013, 2022, ControlsFX
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -26,52 +26,11 @@
  */
 package org.controlsfx;
 
-import fxsampler.FXSamplerProject;
-import javafx.scene.control.Label;
-import javafx.scene.image.ImageView;
-import javafx.scene.layout.StackPane;
-import javafx.scene.layout.VBox;
-import fxsampler.model.WelcomePage;
+import fxsampler.FXSampler;
 
-public class ControlsFXSampler implements FXSamplerProject {
+public class ControlsFXSampler {
 
-    /** {@inheritDoc} */
-    @Override public String getProjectName() {
-        return "ControlsFX";
-    }
-
-    /** {@inheritDoc} */
-    @Override public String getSampleBasePackage() {
-        return "org.controlsfx.samples";
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public String getModuleName() {
-        return "org.controlsfx.controls";
-    }
-
-    /** {@inheritDoc} */
-    @Override public WelcomePage getWelcomePage() {
-        VBox vBox = new VBox();
-        ImageView imgView = new ImageView();
-        imgView.setStyle("-fx-image: url('org/controlsfx/samples/ControlsFX.png');");
-        StackPane pane = new StackPane();
-        pane.setPrefHeight(207);
-        pane.setStyle("-fx-background-image: url('org/controlsfx/samples/bar.png');"
-                + "-fx-background-repeat: repeat-x;");
-        pane.getChildren().add(imgView);
-        Label label = new Label();
-        label.setWrapText(true);
-        StringBuilder desc = new StringBuilder();
-        desc.append("ControlsFX is an open source project for JavaFX that aims ");
-        desc.append("to provide really high quality UI controls and other tools to ");
-        desc.append("complement the core JavaFX distribution.");
-        desc.append("\n\n");
-        desc.append("Explore the available UI controls by clicking on the options to the left.");
-        label.setText(desc.toString());
-        label.setStyle("-fx-font-size: 1.5em; -fx-padding: 20 0 0 5;");
-        vBox.getChildren().addAll(pane, label);
-        return new WelcomePage("Welcome to Controls FX!", vBox);
+    public static void main(String[] args) {
+        FXSampler.main(args);
     }
 }

--- a/controlsfx-samples/src/main/java/org/controlsfx/ControlsFXSamplerProject.java
+++ b/controlsfx-samples/src/main/java/org/controlsfx/ControlsFXSamplerProject.java
@@ -1,0 +1,77 @@
+/**
+ * Copyright (c) 2022, ControlsFX
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *     * Neither the name of ControlsFX, any associated website, nor the
+ * names of its contributors may be used to endorse or promote products
+ * derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL CONTROLSFX BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.controlsfx;
+
+import fxsampler.FXSamplerProject;
+import javafx.scene.control.Label;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.StackPane;
+import javafx.scene.layout.VBox;
+import fxsampler.model.WelcomePage;
+
+public class ControlsFXSamplerProject implements FXSamplerProject {
+
+    /** {@inheritDoc} */
+    @Override public String getProjectName() {
+        return "ControlsFX";
+    }
+
+    /** {@inheritDoc} */
+    @Override public String getSampleBasePackage() {
+        return "org.controlsfx.samples";
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String getModuleName() {
+        return "org.controlsfx.controls";
+    }
+
+    /** {@inheritDoc} */
+    @Override public WelcomePage getWelcomePage() {
+        VBox vBox = new VBox();
+        ImageView imgView = new ImageView();
+        imgView.setStyle("-fx-image: url('org/controlsfx/samples/ControlsFX.png');");
+        StackPane pane = new StackPane();
+        pane.setPrefHeight(207);
+        pane.setStyle("-fx-background-image: url('org/controlsfx/samples/bar.png');"
+                + "-fx-background-repeat: repeat-x;");
+        pane.getChildren().add(imgView);
+        Label label = new Label();
+        label.setWrapText(true);
+        StringBuilder desc = new StringBuilder();
+        desc.append("ControlsFX is an open source project for JavaFX that aims ");
+        desc.append("to provide really high quality UI controls and other tools to ");
+        desc.append("complement the core JavaFX distribution.");
+        desc.append("\n\n");
+        desc.append("Explore the available UI controls by clicking on the options to the left.");
+        label.setText(desc.toString());
+        label.setStyle("-fx-font-size: 1.5em; -fx-padding: 20 0 0 5;");
+        vBox.getChildren().addAll(pane, label);
+        return new WelcomePage("Welcome to Controls FX!", vBox);
+    }
+}

--- a/controlsfx-samples/src/main/resources/META-INF/services/fxsampler.FXSamplerProject
+++ b/controlsfx-samples/src/main/resources/META-INF/services/fxsampler.FXSamplerProject
@@ -1,1 +1,1 @@
-org.controlsfx.ControlsFXSampler
+org.controlsfx.ControlsFXSamplerProject

--- a/controlsfx/build.gradle
+++ b/controlsfx/build.gradle
@@ -10,9 +10,9 @@ ext {
 version = controlsfx_version
 
 dependencies {
-    testCompile 'junit:junit:[4,)'
-    testCompile 'org.testfx:testfx-core:4.0.16-alpha'
-    testRuntime ('org.testfx:openjfx-monocle:jdk-12.0.1+2') {
+    testImplementation 'junit:junit:[4,)'
+    testImplementation 'org.testfx:testfx-core:4.0.16-alpha'
+    testRuntimeOnly ('org.testfx:openjfx-monocle:jdk-12.0.1+2') {
         exclude group: 'org.openjfx'
     }
     testImplementation 'org.hamcrest:hamcrest:2.2'

--- a/controlsfx/build.gradle
+++ b/controlsfx/build.gradle
@@ -90,7 +90,7 @@ javadoc {
                 "javafx.controls/com.sun.javafx.scene.control=org.controlsfx.controls",
                 "javafx.controls/com.sun.javafx.scene.control.behavior=org.controlsfx.controls"
         ]
-)
+    )
     // All doc-files are located in src/main/docs because Gradle's javadoc doesn't copy
     // over the doc-files if they are embedded with the sources. I find this arrangement
     // somewhat cleaner anyway (never was a fan of mixing javadoc files with the sources)

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.0.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.1-bin.zip

--- a/settings.gradle
+++ b/settings.gradle
@@ -13,7 +13,7 @@ buildscript {
         }
     }
     dependencies {
-        classpath 'org.openjfx:javafx-plugin:0.0.7'
+        classpath 'org.openjfx:javafx-plugin:0.0.13'
         classpath 'de.marcphilipp.gradle:nexus-publish-plugin:0.4.0'
     }
 }

